Introduce kvstoremesh, a clustermesh-apiserver companion component

giorio94 opened the pull request:
Please review commit by commit. I'm reporting the description of the main commit here for your convenience:

```
In the current clustermesh implementation, each agent connects to the
kvstore in all remote clusters, pulling information about nodes,
services, identities and IPs. The number of etcd watchers can grow
significantly large (~ 5 * total nodes in the clustermesh), as well
as the number of events to be relayed by etcd to the agents (especially
if the churn rate is high).

kvstoremesh is a new component responsible for synchronizing the
information from the remote kvstores to the local one. In other words,
the local kvstore caches the state of the entire clustermesh, so that
all agents only need to connect to the local kvstore, and not the remote
ones. Overall, this provides better isolation (because the only component
connecting to remote kvstores is now the kvstoremesh operator) and
reduces the number of agents connecting to a single kvstore.

While the number of events to be processed by a single kvstore is
similar to the vanilla clustermesh case, with kvstoremesh the load
is less spiky (because a specific event is propagated to a much lower
number of agents) and it is also easier to rate limit (configuring the
maximum number of keys that can be written by kvstoremesh per second).
This allows to trade latency for reliability in case of high churn in
the clustermesh. Finally, given that the number of watchers is
proportional to the number of nodes in the local cluster only (rather
than the nodes in the entire clustermesh), the dimensioning of the
kvstore is also simplified.

The kvstoremesh concept has been initially proposed by the Cloud-Network
team @Trip.com, led by Arthur Chiao:
http://arthurchiao.art/blog/trip-first-step-towards-cloud-native-security/
```

The Dockerfile and CI modifications required to build the new image, as well as the appropriate extensions to the helm chart, will be introduced in a subsequent PR to limit the size of the current one.

<!-- Description of change -->

```release-note
Introduce kvstoremesh, a clustermesh-apiserver companion component allowing to cache remote cluster information in the local kvstore for increased scalability and separation.
```

giorio94 committed:
clustermesh: add Cached to ClusterConfig capabilities

Extend the CiliumClusterConfig struct with the additional Cached
capability. It conveys the fact that the information concerning the
given cluster is cached from an external kvstore (for instance, by
kvstoremesh). This implies that all keys are stored under the dedicated
`cilium/cache` prefix (rather than the standard `cilium/state` one),
and all are cluster-scoped (including also IPs and identities).

The usage of completely disjoint prefixes is required because currently
nodes and services entries are cluster-scoped, but IPs and identities
ones not. This makes it impossible to watch only the information of a
given cluster if a single kvstore caches entries for multiple clusters.
At the same time, this approach removes the need full downgrade/upgrade
logic which, would be required by a complete switch over (it would be
tricky if not impossible to get right, since old agents would not
understand the new prefixes). Finally, it also clearly highlights that
the information therein stored is relayed from an external source.

For reference, the full "cached" prefixes are:
* cilium/cache/nodes/v1/<cluster-name>/
* cilium/cache/services/v1/<cluster-name>/
* cilium/cache/ip/v1/<cluster-name>/
* cilium/cache/identities/v1/<cluster-name>/

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
clustermesh: introduce the kvstoremesh main logic

In the current clustermesh implementation, each agent connects to the
kvstore in all remote clusters, pulling information about nodes,
services, identities and IPs. The number of etcd watchers can grow
significantly large (~ 5 * total nodes in the clustermesh), as well
as the number of events to be relayed by etcd to the agents (especially
if the churn rate is high).

kvstoremesh is a new component responsible for synchronizing the
information from the remote kvstores to the local one. In other words,
the local kvstore caches the state of the entire clustermesh, so that
all agents only need to connect to the local kvstore, and not the remote
ones. Overall, this provides better isolation (because the only component
connecting to remote kvstores is now the kvstoremesh operator) and
reduces the number of agents connecting to a single kvstore.

While the number of events to be processed by a single kvstore is
similar to the vanilla clustermesh case, with kvstoremesh the load
is less spiky (because a specific event is propagated to a much lower
number of agents) and it is also easier to rate limit (configuring the
maximum number of keys that can be written by kvstoremesh per second).
This allows to trade latency for reliability in case of high churn in
the clustermesh. Finally, given that the number of watchers is
proportional to the number of nodes in the local cluster only (rather
than the nodes in the entire clustermesh), the dimensioning of the
kvstore is also simplified.

The kvstoremesh concept has been initially proposed by the Cloud-Network
team @Trip.com, led by Arthur Chiao:
http://arthurchiao.art/blog/trip-first-step-towards-cloud-native-security/

This commit introduces the logic responsible for watching remote
kvstores and caching the retrieved information in the local one.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
kvstoremesh: add scaffolding

This commit introduces the scaffolding for kvstoremesh, specifically:
* the kvstoremesh main file, which builds the hive and registers the
  appropriate flags. Most notably, it leverages the newly introduced
  kvstoremesh cell and the kvstore cell;
* the metrics server, to expose the information concerning kvstore
  operations;
* the makefile, to build kvstoremesh (inspired from that already
  available for the clustermesh-apiserver).

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
CODEOWNERS: assign kvstoremesh to @cilium/sig-clustermesh

kvstoremesh is a companion component of the already existing
clustermesh-apiserver, hence it feels natural to assign it to the
sig-clustermesh team.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

tommyp1ckles commented on kvstoremesh/metrics/metrics.go:
```diff
@@ -0,0 +1,125 @@
+	}
+
+	mux := http.NewServeMux()
+	mux.Handle("/metrics", promhttp.HandlerFor(mm.registry, promhttp.HandlerOpts{}))
```
The Agent server enables OpenMetrics in HandlerOpts, Im not that familiar with the option but might be something to consider to make sure metrics are consistent among the different components.

giorio94 replied:
I could find `EnableOpenMetrics` being used as an option only for the [Hubble metrics](https://github.com/cilium/cilium/blob/7ec43fe8c6bf53bba6476132200aa1623806713b/pkg/hubble/metrics/metrics.go#L78-L79), but not for the [agent metrics](https://github.com/cilium/cilium/blob/7ec43fe8c6bf53bba6476132200aa1623806713b/pkg/metrics/registry.go#LL79C17-L79C17). Am I missing anything?

tommyp1ckles replied:
Yeah not sure if this is really necessary.

tommyp1ckles commented on kvstoremesh/metrics/metrics.go:
```diff
@@ -0,0 +1,125 @@
+	log.Info("Registering metrics")
+
+	mm.registry.MustRegister(collectors.NewProcessCollector(collectors.ProcessCollectorOpts{}))
+	mm.registry.MustRegister(collectors.NewGoCollector())
```
In the Agent we recently enabled sched latency metrics: https://github.com/cilium/cilium/blob/main/pkg/metrics/registry.go#L116 by default.

The idea was that when people run Cilium in severely CPU constrained environments that we sometimes begin to see strange behaviour as goroutines get large scheduling latency.

Might be worth something to have here too, keep things consistent and we can easily build alerting around that.

giorio94 replied:
Added, thanks for the heads up.

tommyp1ckles commented on kvstoremesh/metrics/metrics.go:
```diff
@@ -0,0 +1,125 @@
+func (mm *metricsManager) Start(hive.HookContext) error {
+	log.Info("Registering metrics")
+
+	mm.registry.MustRegister(collectors.NewProcessCollector(collectors.ProcessCollectorOpts{}))
```
Should these be namespaced (i.e. `cilium_kvstoremesh`)?

giorio94 replied:
It seems that there is currently a bit of inconsistency in the code base, since process-related metrics are currently namespaced in the agents and cilium-operator, and non-namespaced in hubble relay and in the clustermesh-apiserver.  Go related metrics instead are always non-namespaced. Do you feel it would be better change the process ones to be namespaced in clustermesh-apiserver/kvstoremesh?

tommyp1ckles replied:
Good point, the agent/operator metrics declarations also all specify their namespace in-line as well. 
It looks like hubble metrics are just name spaced individually with the "hubble" namespace.

So from what I can tell, with reuising the existing pkg/metrics declarations this will result in using the "cilium_" namespace?

Given that we're reusing the same metric name I think it would make sense for this component to have its own namespace prefix to distinguish between agent/kvstoremesh controller metrics.

giorio94 replied:
>  So from what I can tell, with reuising the existing pkg/metrics declarations this will result in using the "cilium_" namespace?

As for the cilium specific metrics, the [namespace is changed to `cilium_kvstoremesh`](https://github.com/cilium/cilium/blob/96abcb95bc34356c9465de536d35749039419815/kvstoremesh/main.go#L45) (and `cilium_clustermesh_apiserver` in the clustermesh-apiserver case). The only metrics that are currently not namespaced in those two cases are the generic ones (go and process).

tommyp1ckles replied:
ah I see, missed that line, thought this would be the same as the agent. Ok looks good thanks.

giorio94 committed:
docs: add kvstoremesh metrics documentation

This commit extends the metrics documentation with the information about
the newly introduced kvstoremesh metrics. Configuration options will be
added alongside the helm chart modifications.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

tommyp1ckles commented on kvstoremesh/metrics/metrics.go:
```diff
@@ -0,0 +1,129 @@
+	// This is a hack until we can unify this metrics manager with the metrics.Registry.
+	metrics.NewLegacyMetrics()
+	mm.registry.MustRegister(
+		metrics.KVStoreOperationsDuration,
```
Small optional point, it's always nice to add a "version" metrics such as `cilium_version` for a couple reasons:

* Discover ability: low cardinality, generally always enabled metric that can be used to detect running deployments/labels etc.
* Monitoring component versions via Prometheus.

giorio94 replied:
Added, thanks!

zacharysarah commented on Documentation/observability/metrics.rst:
```diff
@@ -179,9 +179,11 @@ metrics conform to the OpenMetrics requirements.
-Cluster Mesh API Server metrics provide insights into both the state of the
-``clustermesh-apiserver`` process and the sidecar etcd instance.
+Cluster Mesh API Server metrics provide insights into the state of the
+``clustermesh-apiserver`` process, the ``kvstoremesh`` process (if enabled)
```
Serial comma nit
```suggestion
``clustermesh-apiserver`` process, the ``kvstoremesh`` process (if enabled),
```

zacharysarah commented on Documentation/observability/metrics.rst:
```diff
@@ -920,3 +922,59 @@ Name                                     Labels
+
+To expose any metrics, invoke ``kvstoremesh`` with the
+``--prometheus-serve-addr`` option. This option takes a ``IP:Port`` pair but
+passing an empty IP (e.g. ``:9964``) will bind the server to all available
```
Use present tense
```suggestion
passing an empty IP (e.g. ``:9964``) binds the server to all available
```

zacharysarah commented on Documentation/observability/metrics.rst:
```diff
@@ -920,3 +922,59 @@ Name                                     Labels
+To expose any metrics, invoke ``kvstoremesh`` with the
+``--prometheus-serve-addr`` option. This option takes a ``IP:Port`` pair but
+passing an empty IP (e.g. ``:9964``) will bind the server to all available
+interfaces (there is usually only one in a container).
```
One what? Please verify.
```suggestion
interfaces (there is usually only one interface in a container).
```

zacharysarah reviewed (approved):
Some small changes required, otherwise LGTM. Approving to unblock with the understanding that changes are required prior to merge.

YutaroHayakawa commented on kvstoremesh/.gitignore:
```diff
@@ -0,0 +1 @@
+kvstoremesh
```
Nit: It's a bit odd to me that the name of the executable is `kvstoremesh` maybe `kvstoremesh-apiserver` or `kvstore-agent`?

giorio94 replied:
Honestly I was doubtful about which name to pick, and I ended up using plain *kvstoremesh* mainly for brevity. I'm personally a bit against `kvstoremesh-apiserver` because it is not actually an apiserver (we already have the same problem with the clustermesh-apiserver, which makes understanding its role a bit difficult). The same applies for `kvstoremesh-operator`, because it is not strictly an operator in the kubernetes sense. I don't have strong opinion though.

YutaroHayakawa reviewed (approved):
LGTM 👍 Beautiful code!

giorio94 committed:
etcd: use default lease TTL value if unset

Currently, the KVstoreLeaseTTL option value is directly used to
initialize the etcd lease manager. Yet, that option is unset when
running the integration tests, causing the etcd server to fallback
to very low TTL values (i.e., 2 seconds), which contribute to test
flakiness. Hence, let's use the default value (15 minutes) when that
option is unset, to prevent issues due to lease expiration.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

tklauser commented on pkg/defaults/defaults.go:
```diff
@@ -26,6 +26,9 @@ const (
 	GopsPortApiserver = 9892
 
+	// GopsPortKVStoreMesh is the default value for option.GopsPort in kvstoremesh
+	GopsPortKVStoreMesh = 9893
```
This value is already used for Hubble Relay's gops port:

https://github.com/cilium/cilium/blob/68b3ae1918b1bd81e2f82d4fb83c896f8fae4f9f/Documentation/operations/system_requirements.rst?plain=1#L431

https://github.com/cilium/cilium/blob/68b3ae1918b1bd81e2f82d4fb83c896f8fae4f9f/pkg/hubble/relay/defaults/defaults.go#L20-L21

I guess this should be 9894 instead.

Would also be nice to document the new value in the table:

https://github.com/cilium/cilium/blob/68b3ae1918b1bd81e2f82d4fb83c896f8fae4f9f/Documentation/operations/system_requirements.rst?plain=1#L413-L436

giorio94 replied:
KVStoreMesh is not running in host network, hence it does not risk conflicting with the ports used on the nodes. Do you think it is better to add it to the system requirements also in that case? (I see that clustermesh-apiserver is present there, although it's unclear to me whether that is intentional or due to historical reasons).

tklauser replied:
I'd still suggest to keep the port values distinct for consistency.

Feel free to omit it from the table, that one admittedly is a bit misleading probably for historical reasons and could use a cleanup.

giorio94 replied:
I've changed the port value to 9894 and opened https://github.com/cilium/cilium/pull/26230 for the cleanup. PTAL

tklauser commented on kvstoremesh/option/config.go:
```diff
@@ -0,0 +1,32 @@
+	PprofAddress = "localhost"
+
+	// PprofPort is the default value for pprof in kvstoremesh
+	PprofPort = 6064
```
Would be nice to document this in the following table along with the other pprof ports:

https://github.com/cilium/cilium/blob/68b3ae1918b1bd81e2f82d4fb83c896f8fae4f9f/Documentation/operations/system_requirements.rst?plain=1#L413-L436

tklauser reviewed (approved):
Thanks!

michi-covalent reviewed (approved):
mostly rubber stamping on behalf of the tophat team. lgtm assuming it's getting reviewed by the actual code owner (sig-clustermesh team) ✅

aditighag reviewed (approved):
@giorio94 It's not obvious if there are any policy related changes in the PR. It looks like sig-policy review was requested after this change - https://github.com/cilium/cilium/compare/9a82196551f3ce7b212c8884c3a5b62895d1d489..d5abdcadbb249cc901ae0511ab332d3346864ee8. 
Let me know if that's not the case.

tommyp1ckles reviewed (approved):
Metrics LGTM 🙏
